include/mbedtls/cipher.h

```diff
@@ -74,9 +90,19 @@ typedef enum {
     MBEDTLS_CIPHER_ID_3DES,
     MBEDTLS_CIPHER_ID_CAMELLIA,
     MBEDTLS_CIPHER_ID_BLOWFISH,
-    MBEDTLS_CIPHER_ID_ARC4,
+    MBEDTLS_CIPHER_ID_ARC4 MBEDTLS_DEPRECATED,
```

Comment:
With gcc 5.4.0 (Ubuntu 16.04 package), I just turned on `MBEDTLS_DEPRECATED_WARNING` in `config.h` and the compilation broke:

```  CC    ccm.c
In file included from ../include/mbedtls/ccm.h:26:0,
                 from ccm.c:39:
../include/mbedtls/cipher.h:70:28: error: expected ‘,’ or ‘}’ before ‘__attribute__’
 #define MBEDTLS_DEPRECATED __attribute__((deprecated))
                            ^
../include/mbedtls/cipher.h:93:28: note: in expansion of macro ‘MBEDTLS_DEPRECATED’
     MBEDTLS_CIPHER_ID_ARC4 MBEDTLS_DEPRECATED,
                            ^
../include/mbedtls/cipher.h:70:28: error: expected ‘,’ or ‘}’ before ‘__attribute__’
 #define MBEDTLS_DEPRECATED __attribute__((deprecated))
                            ^
../include/mbedtls/cipher.h:149:29: note: in expansion of macro ‘MBEDTLS_DEPRECATED’
     MBEDTLS_CIPHER_ARC4_128 MBEDTLS_DEPRECATED,
                             ^
Makefile:161: recipe for target 'ccm.o' failed
make: *** [ccm.o] Error 1
```

GCC supports `__attribute__((deprecated))` for functions, variables and types but not for enum constants.

As a workaround I suggest one of the solutions we've discussed before:

```
typedef enum { ... MBEDTLS_CIPHER_ID_ARC4_DEPRECATED ... } mbedtls_cipher_id_t;
typedef MBEDTLS_DEPRECATED mbedtls_deprecated_cipher_id_t mbedtls_cipher_id_t;
#define MBEDTLS_CIPHER_ID_ARC4 ( (mbedtls_deprecated_cipher_id_t) MBEDTLS_CIPHER_ID_ARC4_DEPRECATED )
```

Reply:
@gilles-peskine-arm I am puzzled, I was sure I had tested this. Will look into it.